### PR TITLE
[no-release-notes] Rowexec refactor

### PR DESCRIPTION
**Pull request by max-hoffman**
Refactor ` Node->RowIter(ctx, row) sql.RowIter` interface method into package visitor. This makes plan nodes slimmer. This also makes it easier to implement alternative execution strategies. A similar package for `Expression.Eval` would be useful.

For example, there is a `sql.NodeExecBuilder` interface now with one `rowexec.builder` implementation that looks something like this:

```golang
func (b *builder) buildNodeExec(ctx *sql.Context, n sql.Node, row sql.Row) (sql.RowIter, error) {
	switch n := n.(type) {
	case *plan.CreateForeignKey:
		return b.buildCreateForeignKey(ctx, n, row)
	case *plan.AlterTableCollation:
		return b.buildAlterTableCollation(ctx, n, row)
	case *plan.CreateRole:
		return b.buildCreateRole(ctx, n, row)
	case *plan.Loop:
		return b.buildLoop(ctx, n, row)
...
```

Every `sql.Node` has a case statement in the builder. Each `b.build{node}` handler is basically `new{node}Iter`, with all of the appropriate code extracted from the `plan` package:

```golang
unc (b *builder) buildLoop(ctx *sql.Context, n *plan.Loop, row sql.Row) (sql.RowIter, error) {
	var blockIter sql.RowIter
	// Currently, acquiring the RowIter will actually run through the loop once, so we abuse this by grabbing the iter
	// only if we're supposed to run through the iter once before evaluating the condition
	if n.OnceBeforeEval {
		var err error
		blockIter, err = b.loopAcquireRowIter(ctx, row, n.Label, n.Block, true)
		if err != nil {
			return nil, err
		}
	}
	iter := &loopIter{
		block:         n.Block,
		label:         strings.ToLower(n.Label),
		condition:     n.Condition,
		once:          sync.Once{},
		blockIter:     blockIter,
		row:           row,
		loopIteration: 0,
	}
	return iter, nil
}
```
Nodes that recursively invoke `RowIter` pass references to the builder. So for example, a join node that previously kept a reference to the join-right `rowIterProvider` now keeps the `sql.Node` and a `sql.NodeExecBuilder` reference when we re-iterate the right relation:

```golang

type crossJoinIterator struct {
	l   sql.RowIter
	r   sql.RowIter
	rp sql.Node
	b  sql.NodeExecBuilder
...

		if i.r == nil {
			iter, err := i.b.Build(ctx, i.rp, i.leftRow)
...
```

All plan execution tests are moved into rowexec.

The expected dependency chain is `parse` -> `plan` -> `rowexec` -> `analyzer`. The sql package cannot depend on `rowexec`.

Subquery expressions are passed a reference to the analyzer's `sql.NodeExecBuilder` during analysis to bridge the Eval->RowIter gap.

### CODE REVIEW

File: sql/core.go [hunk 1]
```diff
@@ -70,7 +71,7 @@ type Node interface {
 	Children() []Node
 	// RowIter produces a row iterator from this node. The current row being evaluated is provided, as well the context
 	// of the query.
-	RowIter(ctx *Context, row Row) (RowIter, error)
+	//RowIter(ctx *Context, row Row) (RowIter, error)
```
Comment (zachmu) [inline sql/core.go:74]: Don't forget to clean this up

**Review by zachmu — COMMENTED** [review]
So I haven't looked at this very closely, just checked out the general idea and a couple implementations of it.

I'm not opposed to this change, but I also don't understand why. What's the rationale here? What does this buy us?

**Comment by max-hoffman** [comment]
> So I haven't looked at this very closely, just checked out the general idea and a couple implementations of it.
> 
> I'm not opposed to this change, but I also don't understand why. What's the rationale here? What does this buy us?

This is geared towards 1) simplifying the code base for non-neckbeards, 2) swapping out the information architecture, and 3) making it easier to test performance refactors.

1/2) In a month I'd like sql.Node to have two methods basically, Children and WithChildren. Moving lifecycles into dedicated packages will just make it more concrete that the IR moves through stages: parse AST -> qualify and resolve symbols into control flow graph -> coerce expression types -> semantic validation of CFG -> normalize into a canonical plan -> explore and cost variations of canonical plan -> execution format.

3) If sql.Node is light, I think it will be easier to layer in memory reuse. Also if we have a bridge point between analysis and exec, it's pretty easy to test alternative expression/node execution formats without replacing all of them in one go. You could imagine moving DML/DDL/triggers into before/after execution stages, and using a vectorized/batching execution format for SELECT nodes with high expected costs. The interface bridges up-front open up optionality.

**Comment by max-hoffman** [comment]
Not a primary goal, but putting these things into distinct packages also makes it easier for multiple people to work on GMS. Significant PRs hopefully won't need to keep touching 100 files.

**Pull request merged**